test/integration/targets/gitlab_project_variable/tasks/main.yml

```diff
@@ -0,0 +1,250 @@
+    api_url: "{{ gitlab_host }}"
+    api_token: "{{ gitlab_login_token }}"
+    project: "{{ gitlab_project_name }}"
+    vars:
+      ACCESS_KEY_ID: checkmode
+  check_mode: yes
+  register: gitlab_project_variable_state
+
+- name: check_mode state must not be changed
+  assert:
+    that:
+      - gitlab_project_variable_state is not changed
+
+- name: apply again the value change from check_mode test
+  gitlab_project_variable:
+    api_url: "{{ gitlab_host }}"
+    api_token: "{{ gitlab_login_token }}"
+    project: "{{ gitlab_project_name }}"
+    vars:
+      ACCESS_KEY_ID: checkmode
+  register: gitlab_project_variable_state
+
+- name: state must not be changed
+  assert:
+    that:
+      - gitlab_project_variable_state is not changed
+
+- name: purge all variables at the beginning
+  gitlab_project_variable:
+    api_url: "{{ gitlab_host }}"
+    api_token: "{{ gitlab_login_token }}"
+    project: "{{ gitlab_project_name }}"
+    purge: True
+
+- name: set two test variables
+  gitlab_project_variable:
+    api_url: "{{ gitlab_host }}"
+    api_token: "{{ gitlab_login_token }}"
+    project: "{{ gitlab_project_name }}"
+    vars:
+      ACCESS_KEY_ID: abc123
+      SECRET_ACCESS_KEY: 321cba
+  register: gitlab_project_variable_state
+
+- name: set two test variables state must be changed
+  assert:
+    that:
+      - gitlab_project_variable_state is changed
+      - gitlab_project_variable_state.project_variable.added|length == 2
+      - gitlab_project_variable_state.project_variable.present|length == 0
+      - gitlab_project_variable_state.project_variable.removed|length == 0
+      - gitlab_project_variable_state.project_variable.updated|length == 0
+
+- name: re-set two test variables
+  gitlab_project_variable:
+    api_url: "{{ gitlab_host }}"
+    api_token: "{{ gitlab_login_token }}"
+    project: "{{ gitlab_project_name }}"
+    vars:
+      ACCESS_KEY_ID: abc123
+      SECRET_ACCESS_KEY: 321cba
+  register: gitlab_project_variable_state
+
+- name: re-set two test variables state must not be changed
+  assert:
+    that:
+      - gitlab_project_variable_state is not changed
+      - gitlab_project_variable_state.project_variable.added|length == 0
+      - gitlab_project_variable_state.project_variable.present|length == 2
+      - gitlab_project_variable_state.project_variable.removed|length == 0
+      - gitlab_project_variable_state.project_variable.updated|length == 0
+
+- name: edit one variable
+  gitlab_project_variable:
+    api_url: "{{ gitlab_host }}"
+    api_token: "{{ gitlab_login_token }}"
+    project: "{{ gitlab_project_name }}"
+    vars:
+      ACCESS_KEY_ID: changed
+    purge: False
+  register: gitlab_project_variable_state
+
+- name: edit one variable state must be changed
+  assert:
+    that:
+      - gitlab_project_variable_state.changed
+      - gitlab_project_variable_state.project_variable.added|length == 0
+      - gitlab_project_variable_state.project_variable.present|length == 0
```

Comment:
Shouldn't this be 1 since `ACCESS_KEY_ID` is still there (unchanged)? And `ACCESS_KEY_ID` already was there as well, so from the description, I would actually expect 2.

Reply:
I guess `present` means "of the variables specified in the module options, the ones which were already there and haven't changed". So `unchanged` would be a better term? Also mention in the description that this only considers variables listed in the options.

Reply:
> Shouldn't this be 1 since ACCESS_KEY_ID is still there (unchanged)? And ACCESS_KEY_ID already was there as well, so from the description, I would actually expect 2.

hmm yeah, interesting. That's wrong. I review it.

Reply:
@felixfontein 

I'm not sure what's the better/more readable variant to fix this issue

```python
    existing_variables = list(filter(None, existing_variables))
    if len(existing_variables) > 0 and purge:
        for item in existing_variables:
            this_gitlab.delete_variable(item)
            change = True
            return_value['removed'].append(item)
    elif len(existing_variables) > 0:
        for item in existing_variables:
            return_value['present'].append(item)
```

vs

```python
    existing_variables = list(filter(None, existing_variables))
    if len(existing_variables) > 0:
        for item in existing_variables:
            if purge:
                this_gitlab.delete_variable(item)
                change = True
                return_value['removed'].append(item)
            else:
                return_value['present'].append(item)
```

Reply:
How about
```
    existing_variables = list(filter(None, existing_variables))
    if purge:
        for item in existing_variables:
            this_gitlab.delete_variable(item)
            change = True
            return_value['removed'].append(item)
    else:
        return_value['present'].extend(existing_variables)
```

Reply:
>  So `unchanged` would be a better term?

I thought `present` is the default naming in ansible

Reply:
`present` is the default naming for a *state*. I would find it ok if this would be the list of variables which are there after the module ran. Or maybe even variables which were there before the module ran. But your definition is something else, so IMO the name should be different.